File: src/guard.py
import math
import json
import pygame
import random
import settings

from entity import Entity

class Guard(Entity):
  def __init__(self, id, config, day_duration, entities):
    random.seed()

    attacker = entities[-1]
    x = random.randint(0, settings.WIDTH)
    y = random.randint(0, settings.HEIGHT)

    behaviour = config["behaviour"]
    if behaviour != "reset" and behaviour != "stroll":
      raise Exception("Guard behaviour {} is not valid".format(behaviour))

    super().__init__(pygame.math.Vector2(x, y), (255, 0, 0))

    self.initial_pos = pygame.math.Vector2(x, y)

    self.name = id
    self.check_time = 0
    self.max_check_time = config["max_check_time"] * 60
    self.min_check_time = config["min_check_time"] * 60
    self.show_radius = settings.TILE_SIZE * config["check_radius"]
    self.check_radius = (self.show_radius)**2

    duration_in_ticks = day_duration * 60 * 60

    self.mission = 0
    self.in_mission = False
    self.mission_period = duration_in_ticks / config["number_of_missions"]
    self.max_mission_time = config["max_mission_time"] * 60
    self.min_mission_time = config["min_mission_time"] * 60

    self.wait_for_mission = random.uniform(0, self.mission_period)

    self.entities = entities
    self.entity_per_check = config["entities_per_check"]
    self.checking_entities = list()

    self.behaviour = behaviour

    self.stroll_speed = config["stroll_speed"]
    self.check_speed = config["check_speed"]
    self.speed = self.stroll_speed * settings.TILE_SIZE

    if settings.DEBUG:
      print("Created guard at ({}, {}) and {}".format(self.pos.x, self.pos.y, self.check_probability))


  def perform_check(self):
    if len(self.checking_entities) > 1:
      self.checking_entities = list()
      return

    possible_checks = list()
    for entity in self.entities:
      if entity.is_being_checked():
        continue

      dist = (self.x() - entity.x())**2 + (self.y() - entity.y())**2
      if dist <= self.check_radius:
        possible_checks.append(entity)

    if len(possible_checks) > 0:
      for i in range(self.entity_per_check):
        if i >= len(possible_checks):
          break

        e = random.choice(possible_checks)
        self.checking_entities.append(e)
        possible_checks.remove(e)

      self.start_check()
      for e in self.checking_entities:
        e.start_check()

      self.check_time = random.uniform(self.min_check_time, self.max_check_time)

  def start_mission(self):
    if not self.in_mission:
      self.in_mission = True
      self.speed = self.check_speed * settings.TILE_SIZE
      self.mission = random.uniform(self.min_mission_time, self.max_mission_time)

  def stop_mission(self):
    if self.in_mission:
      self.in_mission = False
      self.speed = self.stroll_speed * settings.TILE_SIZE
      self.wait_for_mission = self.mission_period

      if self.behaviour == "reset":
        self.reset_pos()

  def reset_pos(self):
    self.pos.x = self.initial_pos.x
    self.pos.y = self.initial_pos.y
    self.constrain(self.speed)

  def move(self, speed):
    if self.wait_for_mission > 0:
      self.wait_for_mission -= 1

      if self.behaviour == "reset":
        return
    else:
      self.start_mission()

    if self.mission > 0:
      self.mission -= 1
    else:
      self.stop_mission()

    if self.check_time > 0:
      self.check_time -= 1
      return

    if self.is_checking:
      self.stop_check()

      for e in self.checking_entities:
        e.stop_check(self.check_radius)

    super().move(self.speed)

    if self.in_mission:
      self.perform_check()

  def update(self):
    super().update()

    # Draw check radius
    if settings.FPS <= 120:
      pygame.draw.circle(self.surface, self.color if self.in_mission else (255, 255, 255), self.pos, self.show_radius, 1)

File: src_cpp/guard.cpp
#include "guard.h"

#include "stdio.h"
#include "helpers.h"

Guard::Guard(uint32_t id, const nlohmann::json& config,
             const std::vector<PMovable>& movables,
             const std::vector<Line>& walls,
             SDL_Renderer* renderer)
    : Movable(0, 0, walls, renderer)
    , mId(id)
    , mMovables(movables)
    , mCheckTime(0)
    , mMissionTime(0)
    , mInMission(false)
{
  // For rand() later
  srand (time(NULL));

  mPos = GetRandomPoint();

  SetColor(255, 0, 0);

  mCheckSpeed = float(config["check_speed"]) * TILE_SIZE;
  mStrollSpeed = float(config["stroll_speed"]) * TILE_SIZE;
  mSpeed = mStrollSpeed;

  auto behaviour = config["behaviour"];
  if (behaviour == "stroll")
    mBehaviour = Behaviour::STROLL;
  else if (behaviour == "reset")
    mBehaviour = Behaviour::RESET;
  else
    throw std::runtime_error("Guard behaviour is invalid");

  mShowRadius = float(config["check_radius"]) * TILE_SIZE;
  mCheckRadius = std::pow(mShowRadius, 2.0);

  float maxCheckTime = float(config["max_check_time"]) * 60;
  float minCheckTime = float(config["min_check_time"]) * 60;
  mRandomCheck = std::make_unique<Randomizer>(minCheckTime, maxCheckTime);

  uint32_t dayInTicks = DAY_LENGTH * 60 * 60;
  mInterMissionPeriod = dayInTicks / float(config["number_of_missions"]);

  float maxMissionTime = float(config["max_mission_time"]) * 60;
  float minMissionTime = float(config["min_mission_time"]) * 60;
  mRandomMission = std::make_unique<Randomizer>(minMissionTime, maxMissionTime);

  Randomizer intermission(0, mInterMissionPeriod / 2);
  mWaitForMissionTime = intermission.Uniform();

  mMovablesPerCheck = uint32_t(config["entities_per_check"]);

  mInitialPos.x = mPos.x;
  mInitialPos.y = mPos.y;
}

Guard::~Guard()
{
}

float Guard::CheckRadius() const
{
  return mCheckRadius;
}

void Guard::Update()
{
  Movable::Update();

  if (HIDDEN)
    return;

  // Uncomment to see radius of detection
  // if (!mInMission)
  //   SDL_SetRenderDrawColor(mRenderer, 255, 255, 255, 255);
  // DrawCircle(mPos.x, mPos.y, mShowRadius);
}

void Guard::Move(const Point& goal)
{
  if (mWaitForMissionTime > 0)
  {
    --mWaitForMissionTime;
    if (mBehaviour == Behaviour::RESET)
      return;
  }
  else
  {
    StartMission();
  }

  if (mMissionTime > 0)
    --mMissionTime;
  else
    StopMission();

  if (mCheckTime > 0)
  {
    --mCheckTime;
    return;
  }

  if (IsChecking())
  {
    StopCheck(mId);
    mBeingChecked.clear();
  }

  // See if we can perform check on every employee is found on the guards way
  if (mInMission)
    PerformCheck();

  mState = State::ACTIVE;
  Movable::Move(goal);

  // Wait a few minutes after position is reach
  // Maybe have coffee and look around
  if (!mInMission && mState == State::IDLE)
    mCheckTime = 120; // 4 minutes
}

void Guard::StartMission()
{
  if (mInMission)
    return;

  mCheckTime = 0;
  mInMission = true;
  mSpeed = mCheckSpeed;
  mMissionTime = mRandomMission->Uniform();
}

void Guard::StopMission()
{
  if (!mInMission)
      return;

  mInMission = false;
  mSpeed = mStrollSpeed;
  mWaitForMissionTime = mInterMissionPeriod;

  if (mBehaviour == Behaviour::RESET)
    ResetPosition();

  // Make sure all entities which were being checked are now free to roam again
  for (auto& m : mMovables)
    m->StopCheck(mId);

  mBeingChecked.clear();
}

void Guard::PerformCheck()
{
  if (!mBeingChecked.empty())
    return;

  std::vector<PMovable> possibleChecks;
  for (auto& e : mMovables)
  {
    if (e->IsChecking())
      continue;

    // Guards only check entities within a certain radius
    if (Distance(mPos.x, mPos.y, e->X(), e->Y()) >= mCheckRadius)
      continue;

    // Make sure we are not checking someone through a wall
    Point min = Raycast(mPos, e->Pos(), mWalls);
    if (min == e->Pos())
      possibleChecks.push_back(e);
  }

  if (possibleChecks.empty())
    return;

  // Of the possible checks, only a few are actually selected
  for (uint32_t i = 0; i < mMovablesPerCheck && !possibleChecks.empty(); ++i)
  {
    int index = rand() % possibleChecks.size();
    mBeingChecked.push_back(possibleChecks.at(index));
    possibleChecks.erase(possibleChecks.begin() + index);
  }

  StartCheck(mId);
  for (auto& e : mBeingChecked)
  {
    if (e->IsChecking())
      continue;

    e->StartCheck(mId);
  }

  // Get time to remain in place/checking
  mCheckTime = mRandomCheck->Uniform();
}

void Guard::ResetPosition()
{
  mPos.x = mInitialPos.x;
  mPos.y = mInitialPos.y;
}

void Guard::RayCast(std::vector<PMovable>& checks)
{
  for (auto iter = checks.begin(); iter < checks.end();)
  {
    Point p((*iter)->X(), (*iter)->Y());
    Point min = Raycast(mPos, p, mWalls);

    if (min == p)
      iter++;
    else
      iter = checks.erase(iter);
  }
}